### PR TITLE
Add VFS support to CephFS backend

### PR DESCRIPTION
**Pull request by xhernandez**

### CODE REVIEW

File: vagrant/ansible/roles/sit.cephfs/templates/smb_share.conf.j2 [hunk 1]
```diff
@@ -1,6 +1,9 @@
 [{{ name }}]
-comment = For samba share of volume {{ path }}
-vfs objects = acl_xattr
+comment = For samba share of subvolume {{ name }}
+vfs objects = acl_xattr{% if config.be.variant == 'vfs' %} ceph
```
Comment (phlogistonjohn) [inline vagrant/ansible/roles/sit.cephfs/templates/smb_share.conf.j2:3]: Minor style nitpick. Having an if block start in the middle of this line but continue over multiple lines looks really awkward to me. Perhaps just an if-else-endif would read better?

**Review by phlogistonjohn — COMMENTED** [review]
Generally, looks like you're heading in the right direction to me.

File: vagrant/ansible/roles/sit.cephfs/templates/smb_share.conf.j2 (hunk 1 above)
Comment (xhernandez) [reply vagrant/ansible/roles/sit.cephfs/templates/smb_share.conf.j2:3]: I'll try to make it prettier.

**Comment by obnoxxx** [comment]
@xhernandez  wrote: 
> I'm not sure what's wrong, but most of the tests fail with error NT_STATUS_NO_MEMORY. Even increasing VM's memory to 4 GiB the error is the same. I'll try to investigate this a bit more.

This might already be a first indication that the cephfs vfs module has the same problem as the gluster module: 

for each share connection (not TCP connection), the entire cephfs client lib is instantiated to establish a ceph connection from userspace.  since the  lib is rather heavyweight, this can lead to memory problems when a lot of share connections are established by a test or other workload.

**Comment by xhernandez** [comment]
> > I'm not sure what's wrong, but most of the tests fail with error NT_STATUS_NO_MEMORY. Even increasing VM's memory to 4 GiB the error is the same. I'll try to investigate this a bit more.
> 
> This might already be a first indication that the cephfs vfs module has the same problem as the gluster module:
> 
> for each share connection (not TCP connection), the entire cephfs client lib is instantiated to establish a ceph connection from userspace. since the lib is rather heavyweight, this can lead to memory problems when a lot of share connections are established by a test or other workload.

I don't understand exactly why, but the problem was that CTDB was checking the presence of the path of the samba shares. Since the path is not physically there, CTDB was not going up and clients were unable to connect. I've disabled this check and "only" 5 tests have failed.

I'm not sure how this translates into memory errors, but now they have disappeared.

**Comment by phlogistonjohn** [comment]
There are parts of the code base that chooses some pretty odd error codes. It was probably due to that.

File: vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml [hunk 2]
```diff
@@ -4,7 +4,29 @@
...
+        policy: targeted
+        state: permissive
+
+    - name: Skip check of share path
+      lineinfile:
+        path: /etc/ctdb/script.options
+        create: true
+        search_string: "CTDB_SAMBA_SKIP_SHARE_CHECK"
+        line: "CTDB_SAMBA_SKIP_SHARE_CHECK=yes"
+        state: present
```
Comment (anoopcs9) [inline vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml:26]: Good catch.

For the default VFS GlusterFS mode we always had `/` and it was not a problem. But with CephFS subvolumes share path can no longer be just `/`. Thanks for investigating the issue.

man [ctdb-script.options(5)](https://ctdb.samba.org/manpages/ctdb-script.options.5.html) also lists an option to have such script-specific options to be defined in a separate file i.e, _50.samba.options_ in this case.

File: vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml [hunk 3]
```diff
@@ -4,7 +4,29 @@
...
+  block:
+    - name: Install CephFS VFS module
+      yum:
+        name: samba-vfs-cephfs
+        state: present
+
+    - name: Set selinux in permissive mode
+      selinux:
+        policy: targeted
+        state: permissive
```
Comment (anoopcs9) [inline vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml:18]: Moving forward we will find ways(policies, allow rules, booleans etc) to make Samba-CephFS integration via VFS module work in default _Enforcing_ mode.

**Review by anoopcs9 — COMMENTED** [review]
Good reorg and clean changes 👏🏻 . You found the right place for backend specific _smb.conf_ share definitions by moving those under respective roles.

Everything looks good to me, thanks.

**Comment by anoopcs9** [comment]
> > > I'm not sure what's wrong, but most of the tests fail with error NT_STATUS_NO_MEMORY. Even increasing VM's memory to 4 GiB the error is the same. I'll try to investigate this a bit more.
> > 
> > 
> > This might already be a first indication that the cephfs vfs module has the same problem as the gluster module:
> > for each share connection (not TCP connection), the entire cephfs client lib is instantiated to establish a ceph connection from userspace. since the lib is rather heavyweight, this can lead to memory problems when a lot of share connections are established by a test or other workload.
> 
> I don't understand exactly why, but the problem was that CTDB was checking the presence of the path of the samba shares. Since the path is not physically there, CTDB was not going up and clients were unable to connect. I've disabled this check and "only" 5 tests have failed.

Look what I found: https://wiki.samba.org/index.php/Samba_CTDB_GlusterFS_Cluster_HowTo#The_file_script.options. People already knew but we haven't had any problems reported with GlusterFS as not many might have used subdir based shares while setting up Samba.

File: vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml (hunk 2 above)
Comment (xhernandez) [reply vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml]: I've updated the PR to use the script specific options, but the path of that file can be considered stable ?

File: vagrant/roles/local.defaults/tasks/main.yml [hunk 4]
```diff
@@ -5,7 +5,8 @@
     dest: ./ansible/config.yml
   vars:
     os: "{{ use_distro | default('centos8') }}"
-    be: "{{ backend | default('glusterfs') }}"
+    be: "{{ (backend | default('glusterfs') | split('.'))[0] }}"
+    variant: "{{ (backend | default('glusterfs') | split('.'))[1] | default('default') }}"
```
[the remark this answers is not in the document]
Comment (xhernandez) [reply vagrant/roles/local.defaults/tasks/main.yml:9]: Done

File: vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml (hunk 2 above)
Comment (xhernandez) [reply vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml]: What's the correct path to add the _.options_ file ?

  - /etc/ctdb/events/legacy/
  - /usr/share/ctdb/events/legacy/

Comment (anoopcs9) [reply vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml]: > What's the correct path to add the _.options_ file ?
> 
> - /etc/ctdb/events/legacy/

This should be it. As the man page says:
...
_options for SCRIPT are placed in a file alongside the script, with a ".script" suffix added_
..

Those legacy events scripts we enable using [ctdb.setup](https://github.com/samba-in-kubernetes/sit-environment/blob/main/vagrant/ansible/roles/ctdb.setup/tasks/main.yml) role appear as symlinks under /etc/ctdb/events/legacy/.

> - /usr/share/ctdb/events/legacy/

 All available event scripts are located here.

File: vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml [hunk 5]
```diff
@@ -4,7 +4,29 @@
...
+
+    - name: Skip check of share path
+      lineinfile:
+        path: /usr/share/ctdb/events/legacy/50.samba.options
```
Comment (anoopcs9) [inline vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml:22]: Proper location would be _/etc/ctdb/events/legacy/50.samba.options_.

**Review by anoopcs9 — CHANGES_REQUESTED** [review]

File: vagrant/ansible/roles/sit.cephfs/templates/smb_share.conf.j2 [hunk 6]
```diff
@@ -0,0 +1,14 @@
+[{{ name }}]
+comment = For samba share of subvolume {{ name }}
+vfs objects = acl_xattr
+{%- if config.be.variant == 'vfs' %} ceph
+ceph: config_file = /etc/ceph/sit.ceph.conf
+ceph: user_id = sit
```
[the remark this answers is not in the document]
Comment (xhernandez) [reply vagrant/ansible/roles/sit.cephfs/templates/smb_share.conf.j2]: Fixed

File: vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml (hunk 2 above)
Comment (xhernandez) [reply vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml]: Done. Thanks

File: vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml (hunk 5 above)
Comment (xhernandez) [reply vagrant/ansible/roles/samba.setup/tasks/cephfs/main.yml]: Done

**Review by anoopcs9 — APPROVED** [review]

**Review by phlogistonjohn — APPROVED** [review]
lgtm

**Pull request merged**